### PR TITLE
#9352 Add new drawing/modify interaction component supports

### PR DESCRIPTION
**Pull request by allyoucanmap**
## Description

This PR implements draw and modify components that could be used for Cesium and OpenLayers. This new components will be used inside the refactor of Annotations

**Please check if the PR fulfills these requirements**
- [x] The commit message follows our guidelines: https://github.com/geosolutions-it/MapStore2/blob/master/CONTRIBUTING.md
- [x] Tests for the changes have been added (for bug fixes / features)
- [x] Docs have been added / updated (for bug fixes / features)

**What kind of change does this PR introduce?** (check one with "x", remove the others)

 - [x] Feature

## Issue

**What is the current behavior?**

#9352

**What is the new behavior?**

Introduced drawing and modify support for Cesium and a new components for OpenLayers

## Breaking change
**Does this PR introduce a breaking change?** (check one with "x", remove the other)
 - [x] No

### CODE REVIEW

File: web/client/components/map/cesium/EditGeoJSONSupport.jsx [hunk 1]
```diff
@@ -0,0 +1,65 @@
...
+ *
+ * This source code is licensed under the BSD-style license found in the
+ * LICENSE file in the root directory of this source tree.
+ */
+
+import { useRef, useEffect } from 'react';
+import CesiumModifyGeoJSONInteraction from '../../../utils/cesium/ModifyGeoJSONInteraction';
+
+/**
+ * Support for 3D drawing, this component provides only the interactions and callback for a drawing workflow
```
Comment (offtherailz) [inline web/client/components/map/cesium/EditGeoJSONSupport.jsx:13]: Explicitly document the fact that this takes only simple geometries, and propeties and add TODOs to allow the impleemntation of multi-geometries for the future.

File: web/client/components/map/cesium/DrawGeometrySupport.jsx [hunk 2]
```diff
@@ -7,268 +7,15 @@
...
-    });
-}
+import CesiumDrawGeometryInteraction from '../../../utils/cesium/DrawGeometryInteraction';
 
 /**
  * Support for 3D drawing, this component provides only the interactions and callback for a drawing workflow
  * @name DrawGeometrySupport
  * @prop {object} map instance of the current map library in use
  * @prop {boolean} active activate the drawing functionalities
  * @prop {string} geometryType type of geometry to draw: Point, LineString or Polygon
```
Comment (offtherailz) [inline web/client/components/map/cesium/DrawGeometrySupport.jsx:17]:
```suggestion
 * @prop {string} geometryType type of geometry to draw. can be: `Point`, `LineString` or `Polygon`
```

File: web/client/components/map/cesium/DrawGeometrySupport.jsx [hunk 3]
```diff
@@ -7,268 +7,15 @@
...
+ * @prop {boolean} geodesic if true the coordinates will be forced to the ellipsoid 0 height
```
Comment (offtherailz) [inline web/client/components/map/cesium/DrawGeometrySupport.jsx:18]:
```suggestion
 * @prop {boolean} geodesic if true the coordinates will be forced to the ellipsoid at 0 height
```

File: web/client/components/map/cesium/DrawGeometrySupport.jsx [hunk 4]
```diff
@@ -279,237 +26,58 @@ function getSampledTerrainPositions(terrainProvider, level = 18, positions) {
  * @prop {boolean} sampleTerrain enable sample terrain functionality only for Point geometry type
  * @prop {number} coordinatesLength number of coordinates expected by a LineString or Polygon geometry, onDrawEnd will be called after the last coordinates added with single click interaction
  * @prop {function} getPositionInfo override the default getPositionInfo function, mainly used for testing
+ * @prop {object} style override the default style of drawing mode (see web/client/utils/DrawUtils.js)
```
Comment (offtherailz) [inline web/client/components/map/cesium/DrawGeometrySupport.jsx:29]:
```suggestion
 * @prop {object} style override the default style of drawing mode (see `web/client/utils/DrawUtils.js`)
```

File: web/client/components/map/cesium/DrawGeometrySupport.jsx [hunk 5]
```diff
@@ -279,237 +26,58 @@ function getSampledTerrainPositions(terrainProvider, level = 18, positions) {
  * @prop {boolean} sampleTerrain enable sample terrain functionality only for Point geometry type
  * @prop {number} coordinatesLength number of coordinates expected by a LineString or Polygon geometry, onDrawEnd will be called after the last coordinates added with single click interaction
```
Comment (offtherailz) [inline web/client/components/map/cesium/DrawGeometrySupport.jsx:27]:
```suggestion
 * @prop {number} coordinatesLength number of coordinates expected by a `LineString` or `Polygon` geometry, `onDrawEnd` will be called after the last coordinates added with single click interaction
```

File: web/client/components/map/cesium/DrawGeometrySupport.jsx [hunk 6]
```diff
@@ -7,268 +7,15 @@
...
 
 /**
  * Support for 3D drawing, this component provides only the interactions and callback for a drawing workflow
  * @name DrawGeometrySupport
  * @prop {object} map instance of the current map library in use
  * @prop {boolean} active activate the drawing functionalities
  * @prop {string} geometryType type of geometry to draw: Point, LineString or Polygon
+ * @prop {boolean} geodesic if true the coordinates will be forced to the ellipsoid 0 height
  * @prop {function} onDrawStart callback triggered at drawing start
  * @prop {function} onDrawing callback triggered at every click/pointerdown events
```
Comment (offtherailz) [inline web/client/components/map/cesium/DrawGeometrySupport.jsx:20]:
```suggestion
 * @prop {function} onDrawing callback triggered at every `click`/`pointerdown` events
```

File: web/client/components/map/cesium/EditGeoJSONSupport.jsx [hunk 7]
```diff
@@ -0,0 +1,65 @@
...
+ * @name EditGeoJSONSupport
+ * @prop {object} map instance of the current map library in use
+ * @prop {boolean} active activate the drawing functionalities
+ * @prop {object} geojson Feature or FeatureCollection GeoJSON data
```
Comment (offtherailz) [inline web/client/components/map/cesium/EditGeoJSONSupport.jsx:17]:
```suggestion
 * @prop {object} geojson `Feature` or `FeatureCollection` GeoJSON data
```

File: web/client/components/map/cesium/EditGeoJSONSupport.jsx [hunk 8]
```diff
@@ -0,0 +1,65 @@
...
+ * @prop {function} onEditEnd triggered one the editing has been completed
+ * @prop {function} toEditProperties convert properties of feature to edit properties geometryType, geodesic and radius are needed to compute the editing. geometryType could be: Point, LineString, Polygon or Circle
```
Comment (offtherailz) [inline web/client/components/map/cesium/EditGeoJSONSupport.jsx:19]:
```suggestion
 * @prop {function} toEditProperties convert properties of feature to edit properties `geometryType`, `geodesic` and `radius` are needed to compute the editing. `geometryType` could be: `Point`, `LineString`, `Polygon` or `Circle`
```

File: web/client/components/map/cesium/EditGeoJSONSupport.jsx [hunk 9]
```diff
@@ -0,0 +1,65 @@
...
+ * @prop {function} fromEditProperties restore properties of the feature to the original one
+ * @prop {number} mouseMoveThrottleTime throttle value to limit the mousemove event callback and improve the interaction, default 100ms
```
Comment (offtherailz) [inline web/client/components/map/cesium/EditGeoJSONSupport.jsx:21]:
```suggestion
 * @prop {number} mouseMoveThrottleTime throttle time in milliseconds, to limit the mousemove event callback and improve the interaction. default value is 100
```

File: web/client/components/map/cesium/EditGeoJSONSupport.jsx [hunk 10]
```diff
@@ -0,0 +1,65 @@
...
+ * @prop {object} style override the default style of drawing mode (see web/client/utils/DrawUtils.js)
```
Comment (offtherailz) [inline web/client/components/map/cesium/EditGeoJSONSupport.jsx:22]:
```suggestion
 * @prop {object} style override the default style of drawing mode (see `web/client/utils/DrawUtils.js`)
```

File: web/client/components/map/openlayers/DrawGeometrySupport.jsx [hunk 11]
```diff
@@ -0,0 +1,67 @@
...
+
+/**
+ * Support for 2D drawing, this component provides only the interactions and callback for a drawing workflow
+ * @name DrawGeometrySupport
+ * @prop {object} map instance of the current map library in use
+ * @prop {boolean} active activate the drawing functionalities
+ * @prop {string} geometryType type of geometry to draw: Circle, Point, LineString or Polygon
+ * @prop {string} geodesic if true the coordinates will visualized as geodetic
+ * @prop {function} onDrawStart callback triggered at drawing start
+ * @prop {function} onDrawing callback triggered at every click/pointerdown events
```
Comment (offtherailz) [inline web/client/components/map/openlayers/DrawGeometrySupport.jsx:20]:
```suggestion
 * @prop {function} onDrawing callback triggered at every `click`/`pointerdown` events
```

File: web/client/components/map/openlayers/DrawGeometrySupport.jsx [hunk 12]
```diff
@@ -0,0 +1,67 @@
...
+ * @prop {function} onDrawEnd callback triggered at drawing end with double click event or single click if coordinatesLength is defined
```
Comment (offtherailz) [inline web/client/components/map/openlayers/DrawGeometrySupport.jsx:21]:
```suggestion
 * @prop {function} onDrawEnd callback triggered at drawing end with double click event or single click if `coordinatesLength` is defined
```

File: web/client/components/map/openlayers/DrawGeometrySupport.jsx [hunk 13]
```diff
@@ -0,0 +1,67 @@
...
+ * @prop {number} coordinatesLength number of coordinates expected by a LineString or Polygon geometry, onDrawEnd will be called after the last coordinates added with single click interaction
+ * @prop {object} style override the default style of drawing mode (see web/client/utils/DrawUtils.js)
```
Comment (offtherailz) [inline web/client/components/map/openlayers/DrawGeometrySupport.jsx:23]:
```suggestion
 * @prop {object} style override the default style of drawing mode (see `web/client/utils/DrawUtils.js`)
```

File: web/client/components/map/openlayers/DrawGeometrySupport.jsx [hunk 14]
```diff
@@ -0,0 +1,67 @@
...
+ * Support for 2D drawing, this component provides only the interactions and callback for a drawing workflow
+ * @name DrawGeometrySupport
+ * @prop {object} map instance of the current map library in use
+ * @prop {boolean} active activate the drawing functionalities
+ * @prop {string} geometryType type of geometry to draw: Circle, Point, LineString or Polygon
+ * @prop {string} geodesic if true the coordinates will visualized as geodetic
+ * @prop {function} onDrawStart callback triggered at drawing start
+ * @prop {function} onDrawing callback triggered at every click/pointerdown events
+ * @prop {function} onDrawEnd callback triggered at drawing end with double click event or single click if coordinatesLength is defined
+ * @prop {number} coordinatesLength number of coordinates expected by a LineString or Polygon geometry, onDrawEnd will be called after the last coordinates added with single click interaction
```
Comment (offtherailz) [inline web/client/components/map/openlayers/DrawGeometrySupport.jsx:22]:
```suggestion
 * @prop {number} coordinatesLength number of coordinates expected by a `LineString` or `Polygon` geometry, `onDrawEnd` will be called after the last coordinates added with single click interaction
```

File: web/client/components/map/openlayers/EditGeoJSONSupport.jsx [hunk 15]
```diff
@@ -0,0 +1,64 @@
...
+ *
+ * This source code is licensed under the BSD-style license found in the
+ * LICENSE file in the root directory of this source tree.
+ */
+
+import { useEffect, useRef } from 'react';
+import OpenLayersModifyGeoJSONInteraction from '../../../utils/openlayers/ModifyGeoJSONInteraction';
+
+/**
+ * Support for 32 drawing, this component provides only the interactions and callback for a drawing workflow
```
[the remark this answers is not in the document]
Comment (offtherailz) [reply web/client/components/map/openlayers/EditGeoJSONSupport.jsx:13]:
```suggestion
 * Support for 2D drawing, this component provides only the interactions and callback for a drawing workflow
```

File: web/client/components/map/openlayers/EditGeoJSONSupport.jsx [hunk 16]
```diff
@@ -0,0 +1,64 @@
...
+ * @name EditGeoJSONSupport
+ * @prop {object} map instance of the current map library in use
+ * @prop {boolean} active activate the drawing functionalities
+ * @prop {object} geojson Feature or FeatureCollection GeoJSON data
```
Comment (offtherailz) [inline web/client/components/map/openlayers/EditGeoJSONSupport.jsx:17]:
```suggestion
 * @prop {object} geojson `Feature` or `FeatureCollection` GeoJSON data
```

File: web/client/components/map/openlayers/EditGeoJSONSupport.jsx [hunk 17]
```diff
@@ -0,0 +1,64 @@
...
+ * @prop {function} onEditEnd triggered one the editing has been completed
+ * @prop {function} toEditProperties convert properties of feature to edit properties geometryType, geodesic and radius are needed to compute the editing. geometryType could be: Point, LineString, Polygon or Circle
```
Comment (offtherailz) [inline web/client/components/map/openlayers/EditGeoJSONSupport.jsx:19]:
```suggestion
 * @prop {function} toEditProperties convert properties of feature to edit properties `geometryType`, `geodesic` and `radius` are needed to compute the editing. `geometryType` could be: `Point`, `LineString`, `Polygon` or `Circle`
```

File: web/client/components/map/openlayers/EditGeoJSONSupport.jsx [hunk 18]
```diff
@@ -0,0 +1,64 @@
...
+ * @prop {function} fromEditProperties restore properties of the feature to the original one
+ * @prop {object} style override the default style of drawing mode (see web/client/utils/DrawUtils.js)
```
Comment (offtherailz) [inline web/client/components/map/openlayers/EditGeoJSONSupport.jsx:21]:
```suggestion
 * @prop {object} style override the default style of drawing mode (see `web/client/utils/DrawUtils.js`)
```

File: web/client/utils/cesium/DrawGeometryInteraction.js [hunk 19]
```diff
@@ -0,0 +1,598 @@
...
+ * @param {boolean} options.depthTestAgainstTerrain apply depthTestAgainstTerrain while drawing inside the default getPositionInfo function
+ * @param {object} options.style style for drawing geometries, see the web/client/DrawUtils.js file
+ * @param {boolean} options.geodesic if true the geometries height will be forced to the ellipsoid 0 height
+ * @param {boolean} options.sampleTerrain enable the possibility to get the point on terrain, it works only with Point type
+ * @param {number} options.mouseMoveThrottleTime change the throttle time to get feedback on mouse move event
+ * @param {function} options.onDrawStart triggered on draw start
+ * @param {function} options.onDrawing triggered while drawing
+ * @param {function} options.onMouseMove triggered while moving the mouse over the globe
+ * @param {function} options.onDrawEnd triggered when the drawing event is completed (double click)
+ */
```
Comment (offtherailz) [inline web/client/utils/cesium/DrawGeometryInteraction.js:208]:
```suggestion
 * Class to manage all the drawing interaction of Cesium library
 * @param {string} options.type type of drawing, one of: `Point`, `LineString`, `Polygon` or `Circle`
 * @param {object} options.map a Cesium map instance
 * @param {number} options.coordinatesLength maximum count of drawing coordinates
 * @param {function} options.getPositionInfo custom function to return info given a position
 * @param {function} options.getObjectsToExcludeOnPick function that returns a list of primitives to exclude when computing the position with the default `getPositionInfo` function
 * @param {boolean} options.depthTestAgainstTerrain apply `depthTestAgainstTerrain` while drawing inside the default `getPositionInfo` function
 * @param {object} options.style style for drawing geometries, see the `web/client/DrawUtils.js` file
 * @param {boolean} options.geodesic if true the geometries height will be forced to the ellipsoid at 0 height
 * @param {boolean} options.sampleTerrain enable the possibility to get the point on terrain, it works only with `Point` type
 * @param {number} options.mouseMoveThrottleTime change the throttle time in milliseconds to get feedback on mouse move event
 * @param {function} options.onDrawStart triggered on draw start
 * @param {function} options.onDrawing triggered while drawing
 * @param {function} options.onMouseMove triggered while moving the mouse over the globe
 * @param {function} options.onDrawEnd triggered when the drawing event is completed (double click)
 */
```

File: web/client/utils/cesium/MathUtils.js [hunk 20]
```diff
@@ -7,6 +7,7 @@
 */
 
 import * as Cesium from 'cesium';
+import earcut from 'earcut';
```
Comment (offtherailz) [inline web/client/utils/cesium/MathUtils.js:10]: Is this lib included? I can not see it package.json

File: web/client/utils/openlayers/DrawGeometryInteraction.js [hunk 21]
```diff
@@ -0,0 +1,305 @@
...
+ * @param {string} options.type type of drawing, one of: Point, LineString, Polygon or Circle
+ * @param {object} options.map a Cesium map instance
+ * @param {number} options.coordinatesLength maximum count of drawing coordinates
+ * @param {object} options.style style for drawing geometries, see the web/client/DrawUtils.js file
+ * @param {boolean} options.geodesic if true the geometries height will be forced to the ellipsoid 0 height
+ * @param {function} options.onDrawStart triggered on draw start
+ * @param {function} options.onDrawing triggered while drawing
+ * @param {function} options.onDrawEnd triggered when the drawing event is completed (double click)
+ */
+class OpenLayersDrawGeometryInteraction {
```
Comment (offtherailz) [inline web/client/utils/openlayers/DrawGeometryInteraction.js:174]:
```suggestion
/**
 * Class to manage all the drawing interaction of OpenLayers library
 * @param {string} options.type type of drawing, one of: `Point`, `LineString`, `Polygon` or `Circle`
 * @param {object} options.map a Cesium map instance
 * @param {number} options.coordinatesLength maximum count of drawing coordinates
 * @param {object} options.style style for drawing geometries, see the `web/client/DrawUtils.js` file
 * @param {boolean} options.geodesic if true the geometries height will be forced to the ellipsoid at 0 height
 * @param {function} options.onDrawStart triggered on draw start
 * @param {function} options.onDrawing triggered while drawing
 * @param {function} options.onDrawEnd triggered when the drawing event is completed (double click)
 */
class OpenLayersDrawGeometryInteraction {
```

**Review by offtherailz — CHANGES_REQUESTED** [review]
Apart from the changes to interfaces already agreed, the only changes regards typos or improvements to doc readability.

File: web/client/utils/cesium/MathUtils.js (hunk 20 above)
Comment (allyoucanmap) [reply web/client/utils/cesium/MathUtils.js:10]: @offtherailz I think it's included by cesium, if you want I can add it in the package.json?

File: web/client/components/map/cesium/EditGeoJSONSupport.jsx (hunk 9 above)
[the remark this answers is not in the document]
Comment (allyoucanmap) [reply web/client/components/map/cesium/EditGeoJSONSupport.jsx]: I removed it from the props so it's not exposed and 100ms by default

File: web/client/components/map/cesium/EditGeoJSONSupport.jsx [hunk 22]
```diff
@@ -10,26 +10,26 @@ import { useRef, useEffect } from 'react';
 import CesiumModifyGeoJSONInteraction from '../../../utils/cesium/ModifyGeoJSONInteraction';
 
 /**
- * Support for 3D drawing, this component provides only the interactions and callback for a drawing workflow
+ * Support for 3D drawing, this component provides only the interactions and callback for a drawing workflow.
+ * At the moment are supported `Feature` or `FeatureCollection` with single geometries, **does not support multi geometry types**.
```
Comment (offtherailz) [inline web/client/components/map/cesium/EditGeoJSONSupport.jsx:14]:
```suggestion
 * At the moment  only `Feature` or `FeatureCollection` with single geometries are supported. **The component does not support multi geometry types**.
```

File: web/client/utils/cesium/ModifyGeoJSONInteraction.js [hunk 23]
```diff
@@ -87,17 +91,22 @@ function updateFeatureCoordinates(feature, updateCallback) {
 
 /**
  * Class to manage all modify interaction of Cesium library given a GeoJSON as input data
+ * At the moment are supported `Feature` or `FeatureCollection` with single geometries, **does not support multi geometry types**.
```
Comment (offtherailz) [inline web/client/utils/cesium/ModifyGeoJSONInteraction.js:94]:
```suggestion
At the moment  only `Feature` or `FeatureCollection` with single geometries are supported. **The component does not support multi geometry types**.
```

File: web/client/utils/openlayers/ModifyGeoJSONInteraction.js [hunk 24]
```diff
@@ -117,23 +121,23 @@ function getColor(color, opacity) {
 
 /**
  * Class to manage all modify interaction of OpenLayers library given a GeoJSON as input data
+ * At the moment are supported `Feature` or `FeatureCollection` with single geometries, **does not support multi geometry types**.
```
Comment (offtherailz) [inline web/client/utils/openlayers/ModifyGeoJSONInteraction.js:124]:
```suggestion
 * At the moment  only `Feature` or `FeatureCollection` with single geometries are supported. **It does not support multi geometry types**.
```

File: web/client/components/map/cesium/EditGeoJSONSupport.jsx [hunk 25]
```diff
@@ -10,26 +10,26 @@ import { useRef, useEffect } from 'react';
...
+ * Following feature properties are used by the edit tool:
+ * - properties.geodesic {boolean} if true enabled geodesic geometries editing
```
Comment (offtherailz) [inline web/client/components/map/cesium/EditGeoJSONSupport.jsx:16]:
```suggestion
 * - properties.geodesic {boolean} if true, enables the geodesic geometries editing
```

**Review by offtherailz — APPROVED** [review]
Nice speed up of the tests. great job

**Pull request merged**